core/src/main/scala/org/bitcoins/core/crypto/MnemonicCode.scala

```diff
@@ -0,0 +1,247 @@
+package org.bitcoins.core.crypto
+
+import java.security.SecureRandom
+
+import org.bitcoins.core.protocol.NetworkElement
+import org.bitcoins.core.util.{CryptoUtil, Factory}
+import scodec.bits.{BitVector, ByteVector}
+
+import scala.annotation.tailrec
+import scala.io.Source
+
+sealed abstract class MnemonicCode extends NetworkElement {
+
+  private val VALID_LENGTHS = Vector(12, 15, 18, 21, 24)
+  require(
+    VALID_LENGTHS.contains(words.length),
+    // $COVERAGE-OFF$
+    s"Number of words must be one of the following: ${VALID_LENGTHS.mkString(", ")} "
+    // $COVERAGE-ON$
+  )
+
+  require({
+    val entropy = toEntropyWithChecksum
+    isEntropyWithChecksumValid(entropy)
+  }, "Entropy checksum is not valid!")
+
+  /**
+    * Checks that the provided entropy has a valid checksum
+    * attached at the end
+    */
+  private[crypto] def isEntropyWithChecksumValid(
+      entropyWithChecksum: BitVector): Boolean = {
+
+    val codeInfo = MnemonicCode.getMnemonicCodeInfo(words)
+    val entropyNoChecksum = entropyWithChecksum.take(codeInfo.entropyBits)
+
+    val hashedEntropy = CryptoUtil.sha256(entropyNoChecksum).bytes.toBitVector
+    val checksum = hashedEntropy.take(codeInfo.checksumLength)
+
+    entropyNoChecksum ++ checksum == entropyWithChecksum
+
+  }
+
+  def words: Vector[String]
+  override def bytes: ByteVector = ByteVector(words.mkString(" ").getBytes)
```

Comment:
Is this right? Is the `words.mkString(" ").getBytes` going to be compatible with 

```
UTF-8 using Normalization Form Compatibility Decomposition (NFKD). 
```